### ISSUE TITLE
Autoninja is incompatible with the instructions in the wiki for using GOMA

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#81510]
If you follow the wiki instructions for getting Fuchsia RBE based GOMA working, `autoninja` from depot_tools does not work.

Autoninja seems to expect GOMA to be running from depot_tools. Unfortunately, it _does_ work on CI, and our some of our test scripts (e.g. run_tests.py, the bash scripts in scenarios_app) expect `autoninja` to work, which can break locally.

We have some team members (myself, @jason-simmons) who have reported just working around this by changing `autoninja` to `ninja` locally. We have others (@chinmaygarde) reporting that it works locally for them using depot_tools goma.

If depot_tools/chromium GOMA now works and we don't need to use Fuchsia RBE, then we should update our wiki and inform team members to do so.

If that's not the case, we need to figure out whether we need to avoid `autoninja` in our scripts, or whether we can get autoninja patched to work for our use case.

/cc @zanderso @bdero @godofredoc

**Comment by flar** [comment]
It looks like the latest version of `autoninja` will make assumptions about your GOMA_DIR. In `depot_tools/autoninja` they try to make sure that if you have a `gomacc` that it can run without encountering problems. They look for this executable in `depot_tools/.cipd_bin/gomacc` which would be the location it would have if you were using the depot_tools version of `goma_ctl` to maintain it, but not the right location if you have your own installation of goma.

So, if you have never setup the depot_tools cipd version of goma, then you will not have that executable in your depot_tools directory and the autoninja script won't bother and will run just fine.

If you have that script, it will run it and bail if there is an error.

If you have that script, but you are using a different installation of GOMA, it will bail even though your GOMA setup might be correct.

The `autoninja.py` program should check your value of GOMA_DIR before it makes such assumptions about whether your goma setup is working.

**Comment by flar** [comment]
The question of whether to switch to the depot_tools cipd version of goma should be independent of fixing the autoninja script to be aware of alternate GOMA_DIR installations.

**Comment by chinmaygarde** [comment]
> We have others (@chinmaygarde) reporting that it works locally for them using depot_tools goma.

I apparently also set `GOMA_DIR` to point to `depot_tools/.cipd_bin`.

We should clarify what the instructions ought to be. Is pointing to Fuchsia RBE really necessary? If folks are having to do workaround locally to use Goma correctly, we should figure out a consistent workflow.

Separately, @flar also suggested having an engine doctor command to check for these sorts of local build issues. Perhaps we could put this in `flutter/tools/doctor`?

**Comment by bdero** [comment]
Ah, I wrote a comment [here](https://github.com/flutter/flutter/issues/77267#issuecomment-839369875) before I noticed this issue. I also got tripped up a bit by the `autoninja` behavior described by @flar (which was introduced by [this commit](https://chromium.googlesource.com/chromium/tools/depot_tools.git/+/e952faee4ce22a7436b2c3bf9e9b9a45fb0014ed%5E%21/#F2)).

I agree that:
1. `autoninja.py` should definitely check if `GOMA_DIR` is set rather than always assuming that `.cipd_bin` is correct.
2. This behavior means the [wiki instructions](https://github.com/flutter/flutter/wiki/Compiling-the-engine#general-compilation-tips) currently lead to a non-working `goma` setup.

As for the Fuchsia RBE, using it is still necessary as `goma.chromium.org` doesn't host the [clang revisions](https://github.com/flutter/engine/blob/9a0aff0667d053be9af95eeef65b65b4881bc48b/DEPS#L501-L523) we build against. And so attempting to build against the Chromium RBE currently results in `compiler binary hash mismatch` errors and local fallback builds for every target (see also b/183555689).

**Comment by bdero** [comment]
Made a quick PR for the autoninja bit: https://chromium-review.googlesource.com/c/chromium/tools/depot_tools/+/2891250/2

**Comment by bdero** [comment]
It landed, so no more `.cipd_bin` trouble: `git remote update; ./update_depot_tools`

Another helpful addition to the wiki that I can think of: Using the Fuchsia RBE can be done by just setting the host/port through environment variables that `goma_ctl` will use:
```
  export GOMA_SERVER_HOST=rbe-prod1.endpoints.fuchsia-infra-goma-prod.cloud.goog
  export GOMA_SERVER_PORT=443
```

AFAICT, that's the only special sauce going on, and so regardless of where developers happen to find goma on their system, it can work. Hopefully that DNS will infrequently/never change.

For example: I didn't have Fuchsia set up when I was first looking through the wiki, so I scrappily did `find . | grep goma_ctl` and happened to find it in `depot_tools/.cipd_bin` -- which, unbeknownst to me, avoided the autoninja failure mode but simultaneously introduced the Fuschia RBE/incompatible hash one 🤦).

**Comment by zanderso** [comment]
@bdero Nice! Is this enough info to update the wiki with and close this issue?

**Comment by bdero** [comment]
Yeah I think so -- I'm gonna try the original instructions once I get back home and see if anything else is out of date + update it.

**Comment by flar** [comment]
I just confirmed the following:

- Looked for `.cipd_bin`, it wasn't there
- Ran `gclient sync` in my engine repo (which creates the `.cipd_bin`)
- Looked for `.cipd_bin`, it now exists
- Ran `autoninja` and it ran fine.

Thanks for the fix!

**Comment by bdero** [comment]
@flar cool, thanks for trying it!

I tried out the instructions in the wiki and the included script looks quite good. I just made some minor tweaks and added blurbs about why the RBE is necessary + a quick fix when running into `compiler binary hash mismatch` errors: [diff](https://github.com/flutter/flutter/wiki/Compiling-the-engine/_compare/8c31988d5ac5907a9d099ff52aae95230ec17966...49bea99e493b8dc5857703ee2b2b3bf0ed9823ac)

I think things are looking pretty good with the Goma situation at this point, so I'll mark this closed. But if I missed anything let's reopen and we can fix it up while it's fresh in the mind.
